docs/design-documents/20221028-dead-letter-queue.md

```diff
@@ -0,0 +1,188 @@
+  stop the pipeline.
+- Allow the user to configure the window and number of messages that need to be
+  nacked to consider the pipeline broken and trigger a stop. Example: if 25 of
+  the last 30 messages are nacked stop the pipeline.
+- Messages pushed to the DLQ should contain metadata describing the error and
+  the source connector.
+
+### Questions
+
+- **Should we allow the user to attach custom metadata to messages pushed to
+  the DLQ?**
+  - We decided to add this as future work in the form of allowing users to
+    attach processors to a DLQ.
+
+### Background
+
+Currently Conduit pipelines fail fast. If a message fails to be processed for
+any reason it is negatively acknowledged which causes the pipeline to stop
+running as soon as possible. In reality it's often the case that we don't want
+to stop the pipeline because of one bad message, instead we want to store it
+somewhere safe and keep the pipeline running. The solution is the support for
+dead letter queues or DLQs.
+
+## Implementation
+
+### DLQ Handlers
+
+Internally we already have a place to implement DLQ handlers
+([here](https://github.com/ConduitIO/conduit/blob/d19379efc04d20d12ab9c80df82a29fcef7e8afd/pkg/pipeline/stream/source_acker.go#L114-L118)).
+We need to decide on the API of the DLQ handler, in this design we propose an
+interface with a method that takes a context, a record and an error and returns
+an error. This should be all we need to write that record to the DLQ.
+
+```go
+type DLQHandler struct {
+    func Write(ctx context.Context, r record.Record, reason error) error
+}
+```
+
+The function `DLQHandler.Write` is expected to synchronously write the record
+to the DLQ and return `nil` if that action succeeded or an error otherwise. The
+synchronous write is needed to ensure the DLQ message actually reached its
+destination, otherwise we need to consider the nack failed and need to stop the
+pipeline.
+
+### Log connector
+
+We should implement a connector that simply logs records as they come in. The
+user should be able to configure which level is used when logging the records.
+
+This is needed because it would be the default connector used for the DLQ.
+
+### DLQ Strategy
+
+We propose to implement a single DLQ strategy that would reroute nacked
+messages to a destination connector plugin. The user needs to be able to define
+the plugin name and its settings.
+
+The user can also configure a stop window and stop threshold.
+
+- `stop_window` defines how many last acks/nacks are monitored in the window
+  that controls if the pipeline should stop. 0 disables the window.
+- `stop_threshold` defines the number of nacks in the window that would cause
+  the pipeline to stop.
+
+The default DLQ plugin should be `builtin:log` with a `stop_window` of 1 and a
+`stop_threshold` of 1. This would result in the same behavior as before the
+introduction of DLQs - the first nacked message would stop the pipeline.
+
+### API
```

Comment:
Will a DLQ be accessible through the API in the same a normal connector is? Also, WDYT about exposing the state of the DLQ (current number of nacked messages in the window)?

Reply:
The current idea is to make the DLQ settings configurable through the pipeline endpoints and make it part of the pipeline config, so no, it wouldn't be accessible as a normal connector.

A second option would be to create another connector type for a DLQ connector, but it didn't feel entirely right, since that's still a destination connector that just acts as the DLQ connector. And I had a problem with attaching the nack window settings (we could inject them in the connector settings with a prefix like `conduit.dlq.*`). Seeing a DLQ being configured as part of the connectors in the pipeline config file also seemed off (I'd like to keep the pipeline config as close to the API representation as possible). At that point it seemed to go down the "hacky" route so I rather went with adding it to the pipeline config itself.

A third option I was thinking of is to separate the DLQ config out into a separate endpoint (e.g. `/pipelines/{id}/dlq`), but that seemed excessive, keeping the DLQ config as part of the pipeline config felt simpler.

> WDYT about exposing the state of the DLQ (current number of nacked messages in the window)?

That's a very good point, it's important to be able to monitor the DLQ. I'll add metrics to the design! 📊

Reply:
> so no, it wouldn't be accessible as a normal connector.

:+1: That makes sense. That also prevents a DLQ from being used as a normal destination, which I think is good.

Reply:
Last thing:) Since DLQ is still a destination a connector, adding a DLQ through the gRPC or HTTP API means that we need to stop a pipeline, add it, and restart the pipeline, correct?